ddl/column_with_index.go

```diff
@@ -0,0 +1,683 @@
+			job.Args = append(job.Args, indexIDs, getPartitionIDs(tblInfo))
+		}
+	default:
+		return ver, errors.Trace(dbterror.ErrInvalidDDLJob.GenWithStackByArgs("table", tblInfo.State))
+	}
+	job.SchemaState = colInfo.State
+	return ver, errors.Trace(err)
+}
+
+// getCurrentReorgIndex returns the current reorging index, if current reorging index not in
+// jobs reorg index array, it will treat it as all index is reorged.
+func getCurrentReorgIndex(job *model.Job, idxInfos []*model.IndexInfo) (*model.IndexInfo, bool) {
+	reorgIdxIDs := job.GetNeedReorgIndexIDs()
+	pos := job.CurrentReorgIndexIdx
+	if pos >= len(reorgIdxIDs) {
+		// Nothing need to reorg just return all done
+		return nil, true
+	}
+	idxID := reorgIdxIDs[pos]
+	for _, idx := range idxInfos {
+		if idx.ID == idxID {
+			return idx, false
+		}
+	}
+	// Not found the index info need reorg just return all done
+	// Actually should not hit here.
+	return nil, true
+}
+
+// updateReorgDone is update the need reorg index array position and check
+// if all indexes is reorged. If not move to next one and reset reorg related data.
+func updateReorgDone(job *model.Job, indexInfo *model.IndexInfo) bool {
+	reorgIdxIDs := job.GetNeedReorgIndexIDs()
+	// Check reorg done index is in job.NeedReorgIndexIDs
+	inList := false
+	for _, idxID := range reorgIdxIDs {
+		if indexInfo.ID == idxID {
+			inList = true
+			break
+		}
+	}
+	if !inList {
+		return false
+	}
+
+	job.CurrentReorgIndexIdx++
+	if job.CurrentReorgIndexIdx < len(reorgIdxIDs) {
+		// Move to next then update reorg related struct
+		newReorgMeta := &model.DDLReorgMeta{
+			SQLMode:       job.ReorgMeta.SQLMode,
+			Warnings:      job.ReorgMeta.Warnings,
+			WarningsCount: job.ReorgMeta.WarningsCount,
+			Location:      job.ReorgMeta.Location,
+			ReorgTp:       job.ReorgMeta.ReorgTp,
+			IsDistReorg:   job.ReorgMeta.IsDistReorg,
+		}
+		// Reset reorg meta and snapshot version
+		job.ReorgMeta = newReorgMeta
+		job.SnapshotVer = 0
+		return false
+	}
+	// All reorg finished
+	return true
+}
+
+// rollback related functions
+
+func getColumnMapFromTable(tblInfo *model.TableInfo) map[string]*model.ColumnInfo {
+	ret := make(map[string]*model.ColumnInfo)
+	for _, colInfo := range tblInfo.Columns {
+		ret[colInfo.Name.L] = colInfo
+	}
+	return ret
+}
+
+// checkDropColumnsNeedReorg returns if the drop column job has composite index covered and index doesn't contains auto increment field.
+func checkDropColumnsNeedReorg(ctx sessionctx.Context, tblInfo *model.TableInfo, colName model.CIStr) (bool, error) {
+	_, cidxInfos := listIndexesWithColumn(colName.L, tblInfo.Indices)
+	if len(cidxInfos) == 0 {
+		return false, nil
+	}
+	colMaps := getColumnMapFromTable(tblInfo)
+
+	for _, idxInfo := range cidxInfos {
+		for _, icol := range idxInfo.Columns {
+			if col, have := colMaps[icol.Name.L]; have && mysql.HasAutoIncrementFlag(col.GetFlag()) {
+				if col.Name.L == colName.L && ctx.GetSessionVars().AllowRemoveAutoInc {
```

Comment:
There has a corner case: If TiDB allow drop auto increment column and the composite index only has this column is auto increment shall we allow this column to be dropped?

For example:
```
create table t (a int auto_increment, b int, c int, index i_ab(a, b));
alter table t drop column a;
```
shall we allow drop the column `a`?

Reply:
I think we should allow drop column `a`. What is your concerned about this case?